bundles/org.openhab.binding.folderwatcher/src/main/java/org/openhab/binding/folderwatcher/internal/handler/FolderWatcherHandler.java

```diff
@@ -0,0 +1,263 @@
+                }
+                reply = ftp.getReplyCode();
+
+                if (!FTPReply.isPositiveCompletion(reply)) {
+                    ftp.disconnect();
+                    logger.debug("FTP server refused connection.");
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                            "FTP server refused connection.");
+                    return;
+                }
+            } catch (IOException e) {
+                if (ftp.isConnected()) {
+                    try {
+                        ftp.disconnect();
+                    } catch (IOException e2) {
+                        logger.debug("Can't disconect: ", e2);
+                    }
+                }
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+                return;
+            }
+            __main: try {
+                if (!ftp.login(config.ftpUsername, config.ftpPassword)) {
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, ftp.getReplyString());
+                    ftp.logout();
+                    break __main;
+                }
+                updateStatus(ThingStatus.ONLINE);
+                if (executionJob != null) {
+                    executionJob.cancel(true);
+                }
+                executionJob = scheduler.scheduleWithFixedDelay(this::refreshFTPFolderInformation,
+                        config.pollInterval + 5, config.pollInterval + 5, TimeUnit.SECONDS);
+                return;
+            } catch (IOException e) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+            }
+            // end main
+            return;
+        }
+    }
+
+    private void refreshFTPFolderInformation() {
+        String ftpRootDir = config.ftpDir;
+        if (ftp != null && ftp.isConnected()) {
+            ftp.enterLocalPassiveMode();
+            try {
+                if (ftpRootDir.endsWith("/")) {
+                    ftpRootDir = ftpRootDir.substring(0, ftpRootDir.length() - 1);
+                }
+                if (!ftpRootDir.startsWith("/")) {
+                    ftpRootDir = "/" + ftpRootDir;
+                }
+                currentFtpListing.clear();
+                currentFtpListing.addAll(listDirectory(ftp, ftpRootDir, "", 0, config.listRecursiveFtp));
+                List<String> diffFtpListing = new ArrayList<>(currentFtpListing);
+                diffFtpListing.removeAll(previousFtpListing);
+                diffFtpListing.forEach(file -> triggerChannel(CHANNEL_FILENAME, file));
+                if (!diffFtpListing.isEmpty()) {
+                    try {
+                        WatcherCommon.saveNewListing(diffFtpListing, currentFtpListingFile);
+                    } catch (IOException e2) {
+                        logger.debug("Can't save new listing into file: {}", e2.getMessage());
```

Comment:
Should this cause the thing to go offline?

Reply:
I was thinking about it. In case writing to file is failed on the next iteration channel will receive same list of new files, so it's not entirely breaking thing, but messing with duplicated channel update. Is it enough to consider that thing should go offline?